protocol/consistencychecks.go

```diff
@@ -42,222 +38,243 @@ type ConsistencyChecks struct {
+func (cc *ConsistencyChecks) HandleResponse(requestType int, msg *Response,
 	uname string, key []byte) ErrorCode {
 	if ErrorResponses[msg.Error] {
 		return msg.Error
 	}
+	switch requestType {
+	case RegistrationType, KeyLookupType:
+		if _, ok := msg.DirectoryResponse.(*DirectoryProof); !ok {
+			return ErrorMalformedDirectoryMessage
+		}
+	default:
+		panic("[coniks] Unknown request type.")
+	}
+	if err := cc.updateSTR(requestType, msg); err != nil {
+		return err.(ErrorCode)
+	}
+	if err := cc.checkConsistency(requestType, msg, uname, key); err != Passed {
+		return err
+	}
+	if err := cc.updateTBs(requestType, msg, uname, key); err != nil {
+		return err.(ErrorCode)
+	}
+	return Passed
+}
 
+func (cc *ConsistencyChecks) updateSTR(requestType int, msg *Response) error {
 	var str *m.SignedTreeRoot
-	var err error
+	switch requestType {
+	case RegistrationType, KeyLookupType:
+		str = msg.DirectoryResponse.(*DirectoryProof).STR
+		// FIXME: check whether the STR was issued on time and whatnot.
+		// Maybe it has something to do w/ #81 and client transitioning between epochs.
+		// Try to verify w/ what's been saved
+		if err := cc.verifySTR(str); err == nil {
+			return nil
+		}
+		// Otherwise, expect that we've enterred a new epoch
+		if err := cc.verifySTRConsistency(cc.SavedSTR, str); err != nil {
+			return err
+		}
+	}
+
+	// And update the saved STR
+	cc.oldSTR = cc.SavedSTR
+	cc.SavedSTR = str
+	return nil
+}
 
+func (cc *ConsistencyChecks) updateTBs(requestType int, msg *Response,
+	uname string, key []byte) error {
+	if !cc.useTBs {
+		return nil
+	}
 	switch requestType {
 	case RegistrationType:
-		str, err = cc.verifyRegistration(requestType, msg, uname, key)
+		df := msg.DirectoryResponse.(*DirectoryProof)
+		if df.AP.ProofType() == m.ProofOfAbsence {
+			if err := cc.verifyReturnedPromise(df, uname, key); err != nil {
+				return err
+			}
+			cc.TBs[uname] = df.TB
+		}
+		return nil
+
 	case KeyLookupType:
-		str, err = cc.verifyKeyLookup(requestType, msg, uname, key)
-	case MonitoringType:
-	case KeyLookupInEpochType:
-	default:
-		panic("[coniks] Unknown request type")
+		df := msg.DirectoryResponse.(*DirectoryProof)
+		ap := df.AP
+		str := df.STR
+		proofType := ap.ProofType()
+		switch {
+		case msg.Error == Success && proofType == m.ProofOfInclusion:
+			if err := cc.verifyFulfilledPromise(uname, str, ap); err != nil {
+				return err
+			}
+
+		case msg.Error == Success && proofType == m.ProofOfAbsence:
+			if err := cc.verifyReturnedPromise(df, uname, key); err != nil {
+				return err
+			}
+			cc.TBs[uname] = df.TB
+		}
 	}
+	return nil
+}
 
-	// update the state
-	if err == PassedWithAProofOfAbsence ||
-		err == PassedWithAProofOfInclusion {
-		cc.SavedSTR = str
+func (cc *ConsistencyChecks) checkConsistency(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+	var err error
+	switch requestType {
+	case RegistrationType:
+		err = cc.verifyRegistration(requestType, msg, uname, key)
```

Comment:
It doesn't look like `verifyRegistration()` or `verifyKeyLookup()` use the `requestType` in any way. Maybe just drop this param?